swagger_server/helpers/db.py

```diff
@@ -0,0 +1,29 @@
+from threading import local
+
+from neo4j import GraphDatabase
+
+
+URI = ""
+ENCRYPTED = True
+
+
+class Database(local):
+    db = None
+
+    def __init__(self):
+        self.driver = GraphDatabase.driver(URI, encrypted=ENCRYPTED)
+
+    def query(self, q):
+        with self.driver.session() as s:
+            result = s.write_transaction(lambda tx: tx.run(q))
+
+        return result
+
+    @classmethod
+    def get(cls):
+        if not cls.db:
+            cls.db = Database()
+        return cls.db
+
+    def close(self):
```

Comment:
This method is currently not called. Please implement `__del__` instead to make sure the connection is closed.

Reply:
wait I implemented it here: https://github.com/iqbal-lab-org/dist-api-prototype/pull/6/files#diff-2954fc7c384956b75f943b6bbdd5e9e4R19

Reply:
> I think cleaning up resource in implicit destructors is not a good practice? Maybe the right solution for this is to call close() when it's needed?

I think it is a better practice to release resource an object acquires during its construction/initialisation in its destruction/finalisation stage. That is what RAII is about: https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization

In comparison, if you rely on the consumer to call `close()` method, there are some potential problems:

- the consumer may forget to call `close()` method and hence cause resource leak.
- the consumer may call `close()` multiple times. Depending on what it does in `close()` method, there could be exceptions thrown when calling the `close()` method on the same object for the second time.
- in our particular use case here, because this is a singleton object, the consumer may use this object again after the `close()` method is called. In this case, the object's db connections have already been closed.

I suggested to use `__del__`, but I had another look and it may not be the best option. I originally thought that it works like a destructor/finaliser. But after reading this document and after resources, it may not be safe and portable option: https://docs.python.org/3/reference/datamodel.html#object.__del__. The reason that it may not be safe is that the method is not guaranteed be called. And also, some other Python implementation may not call it at all. Although it does in CPython which is our target implementation. But I want to make sure it is also portable.

So, I would suggest to implement context manager for this object to guarantee finalisation.

Reply:
ok context manager sounds good